projects/lights_control/src/lights_gpio.c

```diff
@@ -0,0 +1,34 @@
+#include "lights_gpio.h"
+#include "event_queue.h"
+#include "gpio.h"
+#include "gpio_mcu.h"
+#include "lights_config.h"
+#include "lights_events.h"
+#include "status.h"
+
+static StatusCode prv_set_peripherals(uint16_t bitset, GPIOState state, LightsConfig *conf) {
+  const GPIOAddress *addresses = conf->addresses;
+  while (bitset) {
+    uint8_t i = __builtin_ffs(bitset) - 1;  // index of first 1 bit
+    status_ok_or_return(gpio_set_state(&addresses[i], state));
+    bitset &= ~(1 << i);  // bit is read, so we clear it
+  }
+  return STATUS_CODE_OK;
+}
+
+StatusCode lights_gpio_init(LightsConfig *conf) {
+  for (uint8_t i = 0; i < *conf->num_addresses; i++) {
+    status_ok_or_return(gpio_init_pin(&conf->addresses[i], conf->gpio_settings_out));
+  }
+  return STATUS_CODE_OK;
+}
+
+StatusCode lights_gpio_process_event(const Event *e, LightsConfig *conf) {
+  if (e->id >= NUM_LIGHTS_EVENTS) {
```

Comment:
You should probably check if your config contains the event instead. If I passed in `LIGHTS_EVENT_CAN_FAULT`, you would access out-of-bounds memory.

Reply:
ah good point!